### ISSUE TITLE
How to merge into upstream?

### ISSUE THREAD

**Issue by madig** [eellak/gsoc2018-cantarell#3]
Hi!
I was wondering how you're going to integrate your work upstream. Not sure how up-to-date your base Cantarell is.

I'm currently experimenting with converting my sources from Glyphs to UFO (https://gitlab.gnome.org/GNOME/cantarell-fonts/merge_requests/6) but may hold off longer if this is a problem.

**Comment by grautesk** [comment]
It’s the first time I’m working on a collaborative design, so I have no experience yet. How did you handle the integration of Cyrillic? As long as the mutual source is a GLYPHS-file [MergeGlyphs](https://glyphsapp.com/tools/mergeglyphs) seems to be an easy way to proceed to me. You could simply merge the Greek glyph set into your source once ready and convert to UFOs ultimately afterwards.

Upcoming week is the final GSoC evaluation. The Greek addition is pretty much done. I will probably change some details especially in regard to diacritics and finish the kerning. I have other stuff to do in the aftermath of GSoC but I think I should have everything ready in about 2 months. Too long?

**Comment by madig** [comment]
I have no experience with MergeGlyphs, but we could try that. Alternatively, I convert to UFOs and convert yours as well later. That should actually work if you tell me the exact git commit your Glyphs file is based on. At Dalton Maag, UFOs are the source and people convert to Glyphs to work on them. The thing is that producing a font from Glyphs and through fontmake results in slightly different binaries, so I'm still investigating if that difference is significant.

**Comment by madig** [comment]
Any news on merging something?

I still have to work out kinks in the Glyphs -> UFO conversion, so merging into the current Glyphs file is ok to me.

Alexei Vanyashin is doing the rounds on the Cyrillics next month, so having Greek in place might be nice.

PS: I'm going to steal your epsilon for my eopen.

**Comment by grautesk** [comment]
I’d be OK with converting in the near future as I’m on Robofont3 now as well. Talked to @irenevl during AtypI conference and just dropped her an email, maybe she’s got something to add.

I have a little free time upcoming, so I can check my design again for stuff that needs to be fixed. I probably need to get rid of Latin metric keys for my Greek spacing, not convenient for a coop project. I’ll let you know; updating should be done in a week, so Alexei can refer to it next month, if necessary.

**Comment by madig** [comment]
That's great to hear.

Have you tried the ufo2glyphs and glyphs2ufo scripts again of an up-to.date glyphsLib? It should work much better. Remember that when using an up-to-date fontmake and dependencies, you need to insert the following XML snippet into each UFO's lib.plist:
```
<key>com.github.googlei18n.ufo2ft.filters</key>
<array>
  <dict>
    <key>name</key>
    <string>propagateAnchors</key>
    <key>pre</key>
    <true />
  </dict>
</array>
```
Not sure the `pre` thing is necessary, haven't tried. This instructs ufo2ft to propagate anchors almost like Glyphs, i.e. if you are using `a` as a component of `a-cy`, this will make sure anchors defined on `a` are copied to `a-cy` so stuff attaches as expected there.

Edit: I just remembered that there may be a bug in glyphsLib or Glyphs that duplicates this key on every round-trip. Or was it fixed? Guh. Bugs, bugs, bugs.

More magic might possibly be needed if you're doing wacky feature code, but that may be a different story.

Note so self: document all the secret sauce in ufo2ft.

**Comment by irenevlachou** [comment]
Hi, not much to add, fonttools would be my preference but you'll still have to check them again afterwards. But since the sources are still in glyphs, I would add the greek set to the latin sources, so you make use of the keys etc and then invert everything to UFOs. @madig Is the merging going to happen without final kerning etc?

**Comment by madig** [comment]
I can merge whenever you have something to merge, I can wait with actually releasing something :) 

@grautesk would you prefer to kern in Glyphs or Robofont? You could even open a merge request and I only merge it when it's done?

I still have to ask Alexei about his workflow, as he's scheduled to work on Cantarell in November or December. If he's opposed to it, stuff might stay as `.glyphs` a while longer. Though I'd prefer UFOs.

**Comment by irenevlachou** [comment]
For kerning the greek we will need final kerned Latin. Also @grautesk is theoretically and practically done with Cantarell since he has completed his GSOC. This requires a separate arrangement, @davelab6 any ideas? How do we deal with GSOC projects that need extra work before they can be considered ready for publication?

**Comment by davelab6** [comment]
Email me a proposal with description of scope, milestones, schedule dates,
prices (USD)

**Comment by madig** [comment]
What is missing from the Greek that needs Latin kerning?

**Comment by grautesk** [comment]
As far as I understand, Latin (+Cyrillic) is still in progress, right? This includes spacing, which might imply Greek spacing also needs an update once everything is finished. Kerning should be done after design and spacing are finalised. Usually, I’d wait for the Latin to be kerned and use values from that to set up the Greek classes.
 
> @grautesk would you prefer to kern in Glyphs or Robofont? You could even open a merge request and I only merge it when it's done?
> 

Tried Mark Frömberg’s Kernkraft for Glyphs plugin during GSoC as I wanted to use as many open source tools as possible but I’d probably prefer MetricsMachine, which works for both.

**Comment by grautesk** [comment]
Regarding the UFO conversion:
> Have you tried the ufo2glyphs and glyphs2ufo scripts again of an up-to.date glyphsLib? It should work much better.
> 

I haven’t yet. I’d continue using Glyphs for everything related to Cantarell until sources are finally converted. ufo2glyphs creates UFO3s requiring RF3, which I have a license for, now.

**Comment by madig** [comment]
It's still in progress, yes.

Alright, then we shall wait and see.

**Comment by madig** [comment]
Just a FYI: Alexei says he might get time to work on Cantarell in January.

**Comment by madig** [comment]
I just tagged Cantarell v0.200: https://gitlab.gnome.org/GNOME/cantarell-fonts/-/tags/v0.200. I think I can now look at this. I still don't have kerning and I still don't really care (😁), let's see.

**Comment by grautesk** [comment]
Thank you for keeping me up to date. v0.2 is quite some progress. The first release generated from UFOs, right?

I’ve had other projects after finishing my studies in 2019 but would have time to finish the monotonic and polytonic Greek. I think it would be best for me to fork the latest version, add my outlines, update the spacing and do a last quality control with feedback from a native speaker. 

Bringing it up to date to finally merge it shouldn’t be too much work. I’d write a project description to propose this to Google (as @davelab6 suggested in October 2018) and let you know once I receive a feedback.

**Comment by madig** [comment]
Yes, first complete Designspace and UFO pipeline unless I misremember :) Also I think I am one of the first adopters of psautohint 2.0 for hinting CFF2 fonts :grin: 

Yes, please do! Still don't know what to do about kerning though. I am somewhat unhappy with the traditional method of kerning, which leads to file and processing time bloat, but currently cannot use a secret alternative method I have experimented with for my employer :( hm. Also, so far, I am not really missing kerning except maybe for a select few obvious spots like "To" and "AV".

**Comment by madig** [comment]
Any luck so far? :)

**Comment by davelab6** [comment]
> currently cannot use a secret alternative method I have experimented with for my employer

Maybe they would like to sell it to Google ;)

**Comment by madig** [comment]
You know who to contact :grin:

**Comment by sdroege** [comment]
Any updates here? :)

**Comment by coxackie** [comment]
I would also like to inquire about any progress. I use Greek very often in GNOME, and it pains me to see the state the glyphs are right now. It is such a beautiful fonts, and it deserves much better. Is there any way we can help here with the process?

**Comment by madig** [comment]
https://gitlab.gnome.org/GNOME/cantarell-fonts/-/merge_requests/19

**Comment by superuser-does** [comment]
✔ Looks like the issue can be closed, though there's still work to be done:
https://gitlab.gnome.org/GNOME/cantarell-fonts/-/tags
